### ISSUE TITLE
Cocoapods Support

### ISSUE THREAD

**Issue by robertjpayne** [Masonry/Snap#12]
##### TL;DR Snappy will have a compatible podspec soon regardless of Cocoapods real support for Swift

So Cocoapods support looks a ways off, mostly because Xcode still doesn't really support static libs for Swift.

External compiled Swift libraries need two pieces to work properly. The first is obviously the code which is either going to be a static (.a) or dynamic (.dylib, .framework). The second is a companion .swiftmodule for each arch inside the code library.

While Xcode doesn't support compiling static libraries for Swift yet:
- When compiling a .framework the compiler emits intermediate .o files just like Objective-C code.
- When compiling a .framework the compiler emits .swiftmodule files for each arch
- You can combine these .o files into a .a static library
- You can use the .a in conjunction with the .swiftmodule files and a few other things from the output .framework and link it and use it in code completion etc.

I've managed to successfully use the above to get a working static library that can compile into an iOS target and run.

Caveats:
- The podspec will likely come with a pre-compiled .a file. The setup and scripts needed to generate it is beyond the scope of what Cocoapods supports currently.
- The podspec will only be compatible with a single version of Xcode. Because Swift's runtime is bundled in at compile time all external Swift code bases must be compiled with the same version of Xcode to avoid any conflicts.

With a bit of luck I'll have this done tomorrow!

**Comment by mxcl** [comment]
Could this technique be used to provide a `.a` file for users who want to use a library in iOS 7?

I'm thinking without CocoaPods for those users. They drop a `.a` into their project and can have the library since the only other option is for them to manually compile the sources from a submodule.

**Comment by robertjpayne** [comment]
@mxcl it's involved but it theoretically could. The main problem is you need a Swift Module to provide the code completion and documentation etc…

I'm using this script below to find all .swift files in specific Git submodules and compile them into compatible static Swift Modules which can then be linked.


```ruby
##
# This script prepares a set of swiftmodules from source components
##

WORKING_DIR = Dir.pwd
MODULES = {
  "Alamofire" => "Submodules/Alamofire/Source/*.swift",
  "QueryKit" => "Submodules/QueryKit/QueryKit/*.swift",
  "SnapKit" =>  "Submodules/SnapKit/Source/*.swift"
}

##
# DO NOT MODIFY BELOW THIS LINE
##

require "Subprocess"
require "tmpdir"

def generate(file, platform, dst=nil)
  name = File.basename(file).split(".")[0]

  build_path = Dir.mktmpdir
  output_path = (dst) ? File.join(dst, name) : File.expand_path("../#{name}", __FILE__)

  sdks = []
  target = ""
  if platform == :ios
    sdks = {
      "iphoneos" => {"armv7" => "arm", "arm64" => "arm64"},
      "iphonesimulator" => {"i386" => "i386", "x86_64" => "x86_64"}
    }
    target = "apple-ios7.0"
  elsif platform == :osx
    sdks = {
      "macosx" => {"i386" => "i386", "x86_64" => "x86_64"}
    }
    target = "apple-macosx10.10"
  end

  FileUtils.rm_r(build_path) if File.exists?(build_path)
  Dir.mkdir(build_path)

  return if sdks.length == 0

  sdks.each do |sdk, archs|
    sdk_path = Subprocess.check_output(["xcrun", "--show-sdk-path", "--sdk", sdk]).chop

    archs.each do |arch_name, arch_output_name|
      puts "----> Generating files for #{arch_name}"
      module_path = File.join(build_path, "#{arch_output_name}.swiftmodule")
      module_doc_path = File.join(build_path, "#{arch_output_name}.swiftdoc")
      dylib_path = File.join(build_path, "#{arch_output_name}.dylib")
      object_path = File.join(build_path, "#{arch_output_name}.o")
      header_path = File.join(build_path, "#{name}.h")

      Subprocess.check_output([
        "xcrun",
        "swiftc",
        "-sdk", sdk_path,
        "-target", "#{arch_name}-#{target}",
        "-emit-objc-header",
        "-emit-module",
        "-emit-library",
        "-module-name", "#{name}",
        "-module-link-name", "swift#{name}",
        "-g",
        "-j8",
        "-O",
        "-o", dylib_path,
        file
      ])

      FileUtils.mv(File.join(build_path, "#{name}.swiftmodule"), module_path)
      FileUtils.mv(File.join(build_path, "#{name}.swiftdoc"), module_doc_path)

      Subprocess.check_output([
        "xcrun",
        "swiftc",
        "-sdk", sdk_path,
        "-target", "#{arch_name}-#{target}",
        "-module-name", "#{name}",
        "-parse-as-library",
        "-g",
        "-j8",
        "-o", object_path,
        "-O",
        "-c", file
      ])

      Subprocess.check_output([
        "libtool",
        "-static",
        object_path,
        "-o",
        File.join(build_path, "#{arch_output_name}.a"),
      ])

      

    end

  end

  puts "----> Packaging"

  FileUtils.rm_r(output_path) if File.exists?(output_path)

  Dir.mkdir(output_path)
  Dir.mkdir(File.join(output_path, "Headers"))
  Dir.mkdir(File.join(output_path, "#{name}.swiftmodule"))

  Dir["#{build_path}/*.h"].each do |f|
    FileUtils.cp(f, File.join(output_path, "Headers", File.basename(f)))
  end

  Dir["#{build_path}/*.swiftmodule"].each do |f|
    FileUtils.cp(f, File.join(output_path, "#{name}.swiftmodule", File.basename(f)))
  end

  Dir["#{build_path}/*.swiftdoc"].each do |f|
    FileUtils.cp(f, File.join(output_path, "#{name}.swiftmodule", File.basename(f)))
  end

  File.open(File.join(output_path, "module.modulemap"), "w+") do |fh|
    fh << """framework module #{name} {
      header \"#{name}.h\"
      export *
      link \"swift#{name}\"
    }"""
  end

  args = [
    "lipo",
    "-create"
  ]
  Dir["#{build_path}/*.a"].each do |archive|
    args << archive
  end
  args += ["-o", File.join(output_path, "libswift#{name}.a")]

  Subprocess.check_output(args)
  
  puts "----> Finished '#{output_path}'"

rescue => e
  puts "----> Failed #{e.message}"
ensure
  FileUtils.rm_r(build_path)
end

MODULES.each do |name, globs|
  contents = ""
  Dir[File.expand_path(globs, WORKING_DIR)].each do |input_file|
    contents << File.read(File.expand_path(input_file, WORKING_DIR))
    contents << "\n\n"
  end

  tmp_path = Dir.mktmpdir
  input_path = File.join(tmp_path, "#{name}.swift")
  File.open(input_path, "w+") do |fh|
    fh.write(contents)
  end
  puts "*** #{name} ***"
  generate(input_path, :ios, File.expand_path("Modules/", WORKING_DIR))
  FileUtils.rm_r(tmp_path)
end
```

**Comment by robertjpayne** [comment]
@mxcl should say this script above only works for pure swift modules, I couldn't be fussed attempting to fuse Objective-C with Swift though I did find you can cheat and combine the .o symbols in the build directory from a framework and combine them into a static lib and manually piece together the remaining bits.

Once you have a Swift Module you need to only add the containing folder to `Swift Import Paths` in Xcode build settings.
